src/FakeItEasy/Captured.of.TArgument.TCapture.cs

```diff
@@ -0,0 +1,87 @@
+namespace FakeItEasy
+{
+    using System;
+    using System.Collections.Generic;
+    using System.ComponentModel;
+    using System.Diagnostics.CodeAnalysis;
+    using System.Linq;
+
+    /// <summary>
+    /// Holds argument values captured from calls to a faked member.
+    /// </summary>
+    /// <typeparam name="TArgument">The type of the argument.</typeparam>
+    /// <typeparam name="TCapture">The type of the value to capture from the argument.</typeparam>
+    public class Captured<TArgument, TCapture>
+    {
+        private readonly List<TCapture> values = [];
+        private readonly Func<TArgument, TCapture> freezer;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Captured{TArgument, TCapture}"/> class.
+        /// </summary>
+        /// <param name="freezer">
+        ///   Transforms incoming argument values before storing the result.
+        ///   Useful when argument values may be mutated between calls and you want
+        ///   to store a copy that will not be modified.
+        /// </param>
+        internal Captured(Func<TArgument, TCapture> freezer) => this.freezer = Guard.AgainstNull(freezer);
+
+        /// <summary>
+        /// Gets an argument constraint object that will be used to constrain this captured method call argument.
+        /// </summary>
+        public INegatableArgumentConstraintManager<TArgument> That =>
+            ServiceLocator.Resolve<IArgumentConstraintManagerFactory>().Create<TArgument>().IsCapturedTo(this);
+
+        /// <summary>
+        /// Gets a constraint that considers any value of this captured argument as valid.
+        /// </summary>
+        /// <remarks>This is a shortcut for the "Ignored"-property.</remarks>
+        [SuppressMessage(
+            "Microsoft.Naming",
+            "CA1707:IdentifiersShouldNotContainUnderscores",
+            Justification = "But it's kinda cool right?")]
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        [CLSCompliant(false)]
+#pragma warning disable SA1300 // Element must begin with upper-case letter
+        public TArgument _ => this.Ignored;
+#pragma warning restore SA1300 // Element must begin with upper-case letter
+
+        /// <summary>
+        /// Gets a constraint that considers any value of this captured argument as valid.
+        /// </summary>
+        public TArgument Ignored => this.That.Matches(x => true, x => x.Write(nameof(this.Ignored)));
+
+        /// <summary>
+        /// Gets the captured values.
+        /// </summary>
+        /// <remarks>
+        /// Will be empty if no arguments were captured.
+        /// </remarks>
+        public IReadOnlyList<TCapture> Values => this.values;

```

Comment:
It occurs to me that we could improve UX in the case where the `Captured` is created and never used to configure a call and then someone tries to access `Values`. They might get no results and be confused. We _could_ flip a flag once `_`, `Ignored`, or `That` is triggered. If it hasn't been flipped, `Values` could throw.

Maybe it's effort for nothing?
(Also, the analyzers will complain if the property throws what's likely to be a `FakeConfigurationException`.)

Reply:
I think it's fine to just return an empty list.

> (Also, the analyzers will complain if the property throws what's likely to be a `FakeConfigurationException`.)

I don't understand. Can you elaborate?

Reply:
I started going down this road. Made `Values` sometimes throw. When I did, some analyzer. Probably whatever FxCop is now. Complained about throwing the unexpected exception from the property. I gave up, threw it away, later felt bad, and then rose the spectre of throwing again now.

Okay. I just checked:

> warning CA1065: get_Values creates an exception of type FakeConfigurationException, an exception type that should not be raised in a property. If this exception instance might be raised, use a
different exception type, convert this property into a method, or change this property's logic so that it no longer raises an exception. (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1065) [C:\Dev\FakeItEasy\src\FakeItEasy\FakeItEasy.c
sproj::TargetFramework=net6.0]

I quote:

> The following exceptions can be thrown from a property get method:

> 
> * [System.InvalidOperationException](https://learn.microsoft.com/en-us/dotnet/api/system.invalidoperationexception) and all derivatives (including [System.ObjectDisposedException](https://learn.microsoft.com/en-us/dotnet/api/system.objectdisposedexception))

> * [System.NotSupportedException](https://learn.microsoft.com/en-us/dotnet/api/system.notsupportedexception) and all derivatives

> * [System.ArgumentException](https://learn.microsoft.com/en-us/dotnet/api/system.argumentexception) (only from indexed get)

> * [System.Collections.Generic.KeyNotFoundException](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.keynotfoundexception) (only from indexed get)

Reply:
Which it seems we don't. I'm content to leave as empty list.

Reply:
Oh, right. I was confused, I thought you were referring to the FakeItEasy analyzer